File: jambeth/jambeth-bytecode/src/main/java/de/osthus/ambeth/bytecode/abstractobject/ImplementAbstractObjectEnhancementHint.java
package de.osthus.ambeth.bytecode.abstractobject;

import de.osthus.ambeth.bytecode.IEnhancementHint;

/**
 * The context for the {@link ImplementAbstractObjectFactory}
 */
public enum ImplementAbstractObjectEnhancementHint implements IEnhancementHint
{
	ImplementAbstractObjectEnhancementHint;

	@SuppressWarnings("unchecked")
	@Override
	public <T extends IEnhancementHint> T unwrap(Class<T> includedContextType)
	{
		if (ImplementAbstractObjectEnhancementHint.class.isAssignableFrom(includedContextType))
		{
			return (T) this;
		}
		return null;
	}
}

File: ambeth/Ambeth.Bytecode/ambeth/bytecode/abstractobject/ImplementAbstractObjectEnhancementHint.cs
using System;

namespace De.Osthus.Ambeth.Bytecode.AbstractObject
{
    /**
     * The context for the {@link ImplementAbstractObjectFactory}
     */
    public class ImplementAbstractObjectEnhancementHint : IEnhancementHint
    {
        public static readonly ImplementAbstractObjectEnhancementHint INSTANCE = new ImplementAbstractObjectEnhancementHint();

        public Object Unwrap(Type includedContextType)
        {
            if (typeof(ImplementAbstractObjectEnhancementHint).IsAssignableFrom(includedContextType))
            {
                return this;
            }
            return null;
        }

        public T Unwrap<T>() where T : IEnhancementHint
        {
            return (T)Unwrap(typeof(T));
        }
    }
}
